src/main/java/org/logstash/filters/DateFilter.java

```diff
@@ -0,0 +1,118 @@
+/* * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.filters;
+
+import org.joda.time.Instant;
+import org.logstash.Event;
+import org.logstash.ext.JrubyEventExtLibrary.RubyEvent;
+import org.logstash.filters.parser.CasualISO8601Parser;
+import org.logstash.filters.parser.JodaParser;
+import org.logstash.filters.parser.TimestampParser;
+import org.logstash.filters.parser.TimestampParserFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class DateFilter {
+  private final String sourceField;
+  private final String[] tagOnFailure;
+  private RubySuccessHandler successHandler;
+  private final List<ParserExecutor> executors = new ArrayList<>();
+  private final ResultSetter setter;
+
+  public interface RubySuccessHandler {
+    void success(RubyEvent event);
+  }
+
+  public DateFilter(String sourceField, String targetField, List<String> tagOnFailure, RubySuccessHandler successHandler) {
+    this.sourceField = sourceField;
+    this.tagOnFailure = tagOnFailure.toArray(new String[0]);
+    this.successHandler = successHandler;
+    if (targetField.equals("@timestamp")) {
+      this.setter = new TimestampSetter();
```

Comment:
because for, whatever reason, event.setField("@timestamp", value) is not the same as event.setTimestamp(value) in the Java event.

I haven't debugged why, but this is present in Logstash 2.4/5.0, so we need to use this strategy